Thank you @iLeonidze, there are still a couple of problems with the tests  and checks.[ This issue](https://github.com/krrome/docling-hierarchical-pdf/actions/runs/24717127979/job/72304001852?pr=22) you can skip with a "#noqa TRY003" in the row in the test file, since it this not functional code.

In terms of the other pytests failures (hierarchical/hierarchy_builder_metadata.py:122: error: Page? has no attribute "get_textpage" ), it seems like those are unrelated to your PR and might be caused by changes in the pymupdf package. Feel free to look into fixing those as well, but otherwise I can do that.

--- a/hierarchical/hierarchy_builder_metadata.py
+++ b/hierarchical/hierarchy_builder_metadata.py
@@ -114,7 +114,10 @@ class HierarchyBuilderMetadata:
                         add_info["coords"] = this_bbox
                     # sometimes the bookmark still points to the previous page, but the header is at the top of the current page
                     # future todo - instead of this try to use the offset of the bookmark pointer!
-                    for page_here in range(page, min(page + 1, page_count) + 1):
+                    pages = [page]
+                    if page + 1 <= page_count:
+                        pages += [page + 1]
+                    for page_here in pages:
                         if "coords" not in add_info:
                             title_ref = re.sub(r"[^A-Za-z0-9]", "", title)
                             actual_title = ""

--- a/tests/test_metadata_toc_page_bounds.py
+++ b/tests/test_metadata_toc_page_bounds.py
@@ -28,7 +28,7 @@ def test_extract_toc_handles_last_page_without_overflow(monkeypatch):
 
         def __getitem__(self, index):
             if index < 0 or index >= len(self.pages):
-                raise AssertionError(f"Unexpected page access: {index}")
+                raise AssertionError(f"Unexpected page access: {index}")  # noqa: TRY003
             return self.pages[index]
 
         @staticmethod

